Docs: add docs for incremental by partition model kind (#2695)
TobikoData/sqlmesh: docs/concepts/models/model_kinds.md

This kind is a good fit for datasets that have the following traits:

* The dataset's records can be grouped by a partitioning key.
* Each record has a partitioning key associated with it.
* It is appropriate to upsert records, so existing records can be overwritten by new arrivals when their partitioning keys match.
* All existing records associated with a given partitioning key can be removed or overwritten when any new record has the partitioning key value.

The column defining the partitioning key is specified in the model's `MODEL` DDL `partitioned_by` key. This example shows the `MODEL` DDL for an `INCREMENTAL_BY_PARTITION` model whose partition key is the row's value for the `region` column:

```sql linenums="1" hl_lines="4"
MODEL (
  name db.events,

>   kind INCREMENTAL_BY_TIME_RANGE,

Reviewer: ```suggestion
  kind INCREMENTAL_BY_PARTITION,
```

  partitioned_by region,
);
```

Compound partition keys are also supported, such as `region` and `department`:

```sql linenums="1" hl_lines="4"
MODEL (
  name db.events,

>   kind INCREMENTAL_BY_TIME_RANGE,

Reviewer: ```suggestion
  kind INCREMENTAL_BY_PARTITION,
```

  partitioned_by (region, department),
);
```

Date and/or timestamp column expressions are also supported (varies by SQL engine). This BigQuery example's partition key is based on the month each row's `event_date` occurred:

```sql linenums="1" hl_lines="4"
MODEL (
  name db.events,
  kind INCREMENTAL_BY_TIME_RANGE,
  partitioned_by DATETIME_TRUNC(event_date, MONTH)
);
```

**Note**: Partial data [restatement](../plans.md#restatement-plans) is not supported for this model kind, which means that the entire table will be recreated from scratch if restated. This may lead to data loss, so data restatement is disabled for models of this kind by default.

### Materialization strategy

> Depending on the target engine, models of the `INCREMENTAL_BY_TIME_RANGE` kind are materialized using the following strategies:

Reviewer: ```suggestion
Depending on the target engine, models of the `INCREMENTAL_BY_PARTITION` kind are materialized using the following strategies:
```

| Engine     | Strategy                                |
|------------|-----------------------------------------|
| Databricks | REPLACE WHERE by partitioning key       |

> | Spark      | DELETE by partitioning key, then INSERT |

Reviewer: This is actually `INSERT OVERWRITE`.
